compiler/noirc_evaluator/src/ssa/ir/instruction/binary.rs

```diff
@@ -16,10 +16,16 @@ use super::{
 pub(crate) enum BinaryOp {
     /// Addition of lhs + rhs.
     Add,
+    /// Addition of lhs + rhs without an overflow check.
+    UncheckedAdd,
     /// Subtraction of lhs - rhs.
     Sub,
+    /// Subtraction of lhs - rhs without an overflow check.
+    UncheckedSub,
     /// Multiplication of lhs * rhs.
     Mul,
+    /// Multiplication of lhs * rhs without an overflow check.
+    UncheckedMul,
```

Comment:
I was thinking we could add a boolean field to these variants rather than duplicating.

Reply:
That's what I initially did but I ended up going with this option. The reasons are:
1. There are a lot of references to `BinaryOp::Add` (and the others) in SSA tests, so we'd need to change them all (this would be fine, though)
2. Conceptually we could think of them as different operations.
3. Their name would potentially match future methods `unchecked_add`, etc., like in Rust

But I don't have a strong preference, so let me know if you'd prefer a boolean.

Reply:
A point in favor of using a boolean: the compiler will force me to check every existing usage, which right now I had to manually do. But it's a one-time thing, so not sure.

Reply:
Sure, we can always refactor this later if we decide the other option is better so no strong feelings.

Reply:
Actually, it's not a one-time thing because then every time we issue an `Add` instruction the compiler will have us think if we want to do an overflow check or not, so we'd potentially lose some optimizations opportunities if we didn't use a bool. I'll change it.

Reply:
On the other hand we'll have to check every usage in this PR, which might be a good thing or not 😁 

I wonder if pushing to a slice and popping from a slice could use unchecked ops, I don't know if when popping we check if the length is zero somewhere before the subtraction.

Reply:
When popping an element from a slice we can probably use subtraction, it shouldn't go below zero. There is the possibility when pushing that we push more than 2^32 - 1 elements.